sources/web/datalab/polymer/components/datalab-editor/datalab-editor.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Type declarations for CodeMirror
+ */
+
+interface CodeMirrorOptions {
+  value: string,
+  mode: string,
+  lineNumbers: boolean,
+  lineWrapping: boolean,
+  theme: string,
+}
+
+interface CodeMirrorEditor {
+  setOption(name: string, value: string): void;
+  getOption(name: string): void;
+}
+
+declare function CodeMirror (element: HTMLElement, options: CodeMirrorOptions): CodeMirrorEditor;
+
+/**
+ * Editor element for Datalab.
+ * Contains a <datalab-toolbar> element at the top, and a full screen editor
+ * that uses CodeMirror.
+ */
+class DatalabEditorElement extends Polymer.Element {
+
+  /**
+   * Path of the file to load in the editor.
+   */
+  public filePath: string;
+
+  private _editor: CodeMirrorEditor;
+  private _theme: string;
+  private _busy: boolean;
+
+  static get is() { return 'datalab-editor'; }
+
+  static get properties() {
+    return {
+      filePath: {
+        type: String,
+        value: '',
+      },
+      _busy: {
+        type: Boolean,
+        value: false
+      },
+    }
+  }
+
+  ready() {
+    super.ready();
+
+    // Get the theme.
+    SettingsManager.getUserSettingsAsync()
+      .then((settings: common.UserSettings) => {
+        if (settings.theme) {
+          this._theme = settings.theme;
+        }
+      })
+      // Get the file contents, or empty string if no path is specified.
+      .then(() => {
+        if (this.filePath !== '') {
+          this._busy = true;
+          return ApiManager.getJupyterFile(this.filePath, true /*asText*/)
+            .then((file: JupyterFile) => file.content)
+            .catch((e: Error) => {
+              // TODO: Handle error visibly to the user.
+              console.log('Could not load specified file: ', e);
+              return '';
```

Comment:
Looking at the screenshots again, it looks like the filename does not appear anywhere on the page, so as a user I don't know what file I am editing. This could be a problem if I have two tabs open on two files that are similar. Am I missing something? Is the filename available to the user anywhere on that page? It must know it so it can save the file, right?
Given that the filename appears somewhere, let's make sure that, when the file fails to load, the filename attached to the editor window gets cleared. I want to make sure we won't fail to load the file, present an empty editing screen, then write that back to the original file when the user says Save.